p4-16/psa/examples/psa-example-recirculate.p4

```diff
@@ -0,0 +1,208 @@
+
+/*
+Copyright 2017 Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <core.p4>
+#include "../psa.p4"
+
+typedef bit<48>  EthernetAddress;
+
+header ethernet_t {
+    EthernetAddress dstAddr;
+    EthernetAddress srcAddr;
+    bit<16>         etherType;
+}
+
+header ipv4_t {
+    bit<4>  version;
+    bit<4>  ihl;
+    bit<8>  diffserv;
+    bit<16> totalLen;
+    bit<16> identification;
+    bit<3>  flags;
+    bit<13> fragOffset;
+    bit<8>  ttl;
+    bit<8>  protocol;
+    bit<16> hdrChecksum;
+    bit<32> srcAddr;
+    bit<32> dstAddr;
+}
+
+struct fwd_metadata_t {
+    bit<32> outport;
+}
+
+struct empty_metadata_t {}
+
+header recirc_metadata_t {
+    bit<8> custom_field;
+}
+
+struct metadata {
+    fwd_metadata_t fwd_metadata;
+    bit<3> custom_clone_id;
+    recirc_metadata_t recirc_header;
+}
+
+struct headers {
+    ethernet_t       ethernet;
+    ipv4_t           ipv4;
+}
+
+parser CommonParser(packet_in buffer,
+                    out headers parsed_hdr,
+                    inout metadata user_meta)
+{
+    state start {
+        transition parse_ethernet;
+    }
+    state parse_ethernet {
+        buffer.extract(parsed_hdr.ethernet);
+        transition select(parsed_hdr.ethernet.etherType) {
+            0x0800: parse_ipv4;
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        buffer.extract(parsed_hdr.ipv4);
+        transition accept;
+    }
+}
+
+parser IngressParserImpl(
+    packet_in buffer,
+    out headers parsed_hdr,
+    inout metadata user_meta,
+    in psa_ingress_parser_input_metadata_t istd,
+    in empty_metadata_t resub_meta,
+    in recirc_metadata_t recirc_meta,
+    out psa_parser_output_metadata_t ostd)
+{
+    CommonParser() p;
+
+    state start {
+        transition select(istd.packet_path) {
+           PacketPath_t.RECIRCULATE: parse_recirc_header;
+           PacketPath_t.NORMAL: parse_ethernet;
+        }
+    }
+
+    state parse_recirc_header {
+        user_meta.recirc_header = recirc_meta;
+	transition accept;
+    }
+
+    state parse_ethernet {
+        p.apply(buffer, parsed_hdr, user_meta);
+	transition accept;
+    }
+}
+
+control ingress(inout headers hdr,
+                inout metadata user_meta,
+                in  psa_ingress_input_metadata_t  istd,
+                inout psa_ingress_output_metadata_t ostd)
+{
+    action do_recirc (PortId_t port) {
+        ostd.egress_port = PORT_RECIRCULATE;
```

Comment:
I think it would be better to use "send_to_port(PORT_RECIRCULATE);" here, as that also sets the drop flag to false.  As written now, the packet will be dropped after ingress because by default the drop flag is true when ingress begins.